Adding PCA9685 controller to Led

frxnz opened the pull request:
Adds support for the [Adafruit PWM Driver](https://www.adafruit.com/products/815) breakout and shield.

Also refactored `Led` to follow the Controller model.

frxnz commented on test/led.js:
```diff
@@ -487,6 +489,297 @@ exports["Led - PWM (Analog)"] = {
+    test.ok(this.spy.calledWith(64, [6, 0, 0, 4095, 15]));
+
+    // this.led.brightness(100);
+    // test.ok(this.spy.calledWith(64, [6, 0, 0, 4095 * 100 / 255, 15]));
```
I was having trouble getting this to pass. The values being passed are correct, but the test fails. I'm wondering if there's a rounding issue.

frxnz replied:
Derp, nevermind. The last element of the array is wrong.

frxnz commented on eg/led-PCA9685.js:
```diff
@@ -0,0 +1,24 @@
+    controller: "PCA9685"
+  });
+
+  // address: The address of the shield.
```
This all needs to be cleaned up

rwaldron replied:
Since `0x40` is the default, you can remove the property from this constructor opts

frxnz commented:
@rwaldron Do I need to regenerate docs or does that get done on merge?

dtex commented on lib/led.js:
```diff
@@ -23,27 +143,36 @@ function Led(opts) {
+    controller = Controllers["Standard"];
+  }
+
+  Object.defineProperties(this, controller);
```
This seems unnecessary since opts alway exists either as a number or an object. If opts does not exist we don't have a pin number so the instance is doomed anyway.

Are their reasons for doing it this way that I don't see?

It's handled like this in Servo and Motor:
````
if (this.opts.controller) {
  controller = typeof this.opts.controller === "string" ?
    Controllers[this.opts.controller] : this.opts.controller;

  Object.defineProperties(this, controller);
}
````

Is that something we should change?

frxnz replied:
@dtex `Led` can be initialized with no parameters, since there is a `defaultLed` hardcoded to pin 13. I can definitely see the reasoning behind this, since Arduino has the LED on pin 13, and it makes the "hello world" dead simple.

dtex replied:
See, I definitely can't speak to LED :-)

Would it be simpler have a default opts object and then decorate that with the user defined opts? That would simplify this and the defaultLed stuff.

frxnz replied:
Yep, that makes a lot of sense. There might still need to be some type checking, but I think having a default object would be more intuitive/readable.

rwaldron replied:
None of this is necessary if you move the relevant parts of this: 

```js
  if (!opts || !opts.hasOwnProperty("controller")) {
    controller = Controllers["Standard"];
  }
  else if (typeof opts.controller === "string") {
    controller = Controllers[opts.controller];
  }
  else {
    controller = opts.controller;
  }

  if (controller == null) {
    controller = Controllers["Standard"];
  }

  Object.defineProperties(this, controller);
```

To some point after this: 

```js
  Board.Component.call(
    this, opts = Board.Options(opts)
  );
```
 

If you look in accelerometer.js, it's correct: 

```js
  var controller = null;
  var err = null;
  
  // ...
  
  Board.Component.call(
    this, opts = Board.Options(opts)
  );

  if (opts.controller && typeof opts.controller === "string") {
    controller = Controllers[opts.controller.toUpperCase()];
  } else {
    controller = opts.controller;
  }

  if (controller == null) {
    controller = Controllers["ANALOG"];
  }

  Object.defineProperties(this, controller);
```

dtex commented:
I can't speak to LED in general, but the Controller implementation looks spot on (save my one question above). @Resseguie and @rwaldron know this class much better than I so I'm sure their input will be forthcoming.

dtex commented:
Oh, and the commits should be squashed. There is so much going on with J5 that having new functionality spread out across multiple commits makes it hard to keep track.

rwaldron commented on eg/led-rgb-PCA9685.js:
```diff
@@ -0,0 +1,28 @@
+    pins: [0, 1, 2],
+    address: 0x40,
+    controller: "PCA9685"
+  });
```
Oh rad, I like this alot :)

rwaldron commented on lib/led.js:
```diff
@@ -1,9 +1,129 @@
+          };
+
+          // Reset
+          this.io.sendI2CWriteRequest(this.address, [this.COMMANDS.PCA9685_MODE1, 0x0]);
```
Update these i2c calls to the new API: https://github.com/jgautier/firmata#i2c (these should be `i2cWriteReg(address, reg, value)`)

rwaldron commented on lib/led.js:
```diff
@@ -1,9 +1,129 @@
+        on = 0;
+        off = this.pwmRange[1] * value / 255;
+
+        this.io.sendI2CWriteRequest(this.address, [this.COMMANDS.LED0_ON_L + 4 * (this.pin), on, on >> 8, off, off >> 8]);
```
`i2cWrite(address, [...bytes])`

rwaldron commented on lib/led.js:
```diff
@@ -1,9 +1,129 @@
+      }
+    }
+  },
+  Standard: {
```
"Standard" => "DEFAULT"

rwaldron commented on lib/led.js:
```diff
@@ -1,9 +1,129 @@
+
+      }
+    },
+    ledWrite: {
```
just call this `write`

rwaldron commented on test/led.js:
```diff
@@ -361,14 +361,16 @@ exports["Led - PWM (Analog)"] = {
     this.led.strobe();
-    // pre 0.8.13, this would change to 1
+    this.clock.tick(1001);
     test.equal(this.led.mode, 3);
```
I don't see why these tests needed clock ticks? They don't test anything that is time-bound, the mode will be updated synchronously.

frxnz replied:
@rwaldron Since the PWM check and `this.mode` reset is now in `ledWrite` (since it's controller-specific), the mode doesn't get reset until the first write has fired.

That said the default delay is actually 10ms, so they should be `this.clock.tick(10);`

rwaldron replied:
I get it now. I'm not sure I like it, but I'm not going to block on that either.

frxnz replied:
I hear you, I was thinking "this isn't the right way to do this" as I was
doing it. I'll give it some more thought and see if I can come up with a
better solution.

On Wed, Jan 21, 2015 at 3:12 PM, Rick Waldron <notifications@github.com>
wrote:

> In test/led.js
> <https://github.com/rwaldron/johnny-five/pull/586#discussion_r23330317>:
>
> >      test.equal(this.led.mode, 3);
> >
> >      this.led.strobe();
> > -    // pre 0.8.13, this would change to 1
> > +    this.clock.tick(1001);
> >      test.equal(this.led.mode, 3);
>
> I get it now. I'm not sure I like it, but I'm not going to block on that
> either.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rwaldron/johnny-five/pull/586/files#r23330317>.
>

rwaldron replied:
Cool, I appreciate that. 

I'm looking forward to this support landing :)

rwaldron commented:
Needs a breadboard diagram to accompany `eg/led-PCA9685.js` and `eg/led-rgb-PCA9685.js`. Adafruit has fritzing components for the breakout, which is sufficient

frxnz commented:
- [x] Updated i2c calls to the new API
- [x] Since 0x40 is the default, you can remove the property from this constructor opts
- [x] "Standard" => "DEFAULT"
- [x] `ledWrite` => `write`
- [x] `this.mode` is read only, eliminates need for repeated PWM checking
- [x] Added breadboard diagram to docs

rwaldron commented:
I'm soooooo excited for this. I just want to play around with it first, then I'll land it :)

rwaldron commented:
Landed https://github.com/rwaldron/johnny-five/commit/a662c55aedbc81dbb8d938cea0970056d4928c69
